### PR TITLE
Adapt `NetworkPolicy`s for components in shoot namespaces

### PR DESCRIPTION
**Pull request by rfranzke**
**How to categorize this PR?**

/area dev-productivity ops-productivity networking security
/kind enhancement

**What this PR does / why we need it**:
The new `NetworkPolicy` controller was enabled for shoot namespaces with #7484. This PR adapts most of the custom policies and makes use of those created by the controller. Note that most cross-namespace policies are not yet covered and will be tackled separately once the controller gets activated for other namespaces in the seed as well.

**Which issue(s) this PR fixes**:
Part of #7352

**Special notes for your reviewer**:
/cc @timuthy @ScheererJ
FYI @istvanballok @rickardsjp @vlvasilev 

**Release note**:

```breaking dependency
Extensions which deploy components to shoot namespaces need to adapt to the new `NetworkPolicy`s. Concretely, the following labels related to `NetworkPolicies` are deprecated and should be replaced:
- `networking.gardener.cloud/to-shoot-apiserver=allowed`, replace it with `networking.resources.gardener.cloud/to-kube-apiserver-tcp-443=allowed`.
- `networking.gardener.cloud/from-shoot-apiserver=allowed`, replace it with the label `networking.resources.gardener.cloud/to-<service-name>-tcp-<container-port>=allowed` on `kube-apiserver` pods.
```
```breaking operator
Before upgrading to this Gardener version, `Seed`s using `.spec.dns.ingressDomain` must now finally be switched to using `.spec.ingress` and `.spec.dns.provider` (as changed with https://github.com/gardener/gardener/pull/3131 back in 2020). Please find more information about it [here](https://github.com/gardener/gardener/tree/master/docs/deployment/deploy_gardenlet_manually.md#kubernetes-cluster-that-should-be-registered-as-a-seed-cluster). The `.spec.dns.ingressDomain` field is deprecated since more than 2 years and will be removed in a future version.
```

### CODE REVIEW

File: pkg/operation/botanist/component/kubeapiserver/deployment.go [hunk 1]
```diff
@@ -714,12 +715,19 @@ func (k *kubeAPIServer) handleVPNSettings(
...
 		k.handleVPNSettingsHA(deployment, secretCAVPN, secretHAVPNSeedClient, secretHAVPNSeedClientSeedTLSAuth)
+		portVPNSeedServer = vpnseedserver.OpenVPNPort
 	} else {
 		k.handleVPNSettingsNonHA(deployment, secretCAVPN, secretHTTPProxy, configMapEgressSelector)
+		portVPNSeedServer = vpnseedserver.EnvoyPort
 	}
 
+	deployment.Spec.Template.Labels = utils.MergeStringMaps(deployment.Spec.Template.Labels, map[string]string{
+		gardenerutils.NetworkPolicyLabel(vpnseedserver.ServiceName, portVPNSeedServer): v1beta1constants.LabelNetworkPolicyAllowed,
+	})
```
Comment (timuthy) [inline pkg/operation/botanist/component/kubeapiserver/deployment.go:751]: We cloud check if `k.values.VPN.Enabled` here.

**Comment by rfranzke** [comment]
I'm not sure why the upgrade tests fail - I tried to reproduce this locally, but everything worked as expected. I'll need to inspect the Prow execution pod to get to know more, I believe...

/test pull-gardener-e2e-kind-upgrade

File: pkg/apis/core/v1beta1/constants/types_constants.go [hunk 2]
```diff
@@ -470,6 +470,9 @@ const (
 	// which should be scraped by Prometheus.
 	// See https://github.com/gardener/gardener/blob/master/docs/concepts/resource-manager.md#overwriting-the-pod-selector-label.
 	LabelNetworkPolicyScrapeTargets = "all-scrape-targets"
+	// LabelNetworkPolicyShootNamespaceAlias is constant for the alias for shoot namespaces used in NetworkPolicy
```
Comment (timuthy) [inline pkg/apis/core/v1beta1/constants/types_constants.go]:
```suggestion
	// LabelNetworkPolicyShootNamespaceAlias is a constant for the alias for shoot namespaces used in NetworkPolicy
```

File: pkg/operation/botanist/component/vpnseedserver/vpn_seed_server.go [hunk 3]
```diff
@@ -713,6 +712,8 @@ func (v *vpnSeedServer) deployService(ctx context.Context, idx *int) error {
 
 	_, err := controllerutils.GetAndCreateOrMergePatch(ctx, v.client, service, func() error {
 		metav1.SetMetaDataAnnotation(&service.ObjectMeta, "networking.istio.io/exportTo", "*")
+		metav1.SetMetaDataAnnotation(&service.ObjectMeta, resourcesv1alpha1.NetworkingNamespaceSelectors, fmt.Sprintf(`[{"matchLabels":{"%s":"%s"}}]`, v1beta1constants.GardenRole, v1beta1constants.GardenRoleIstioIngress))
```
Comment (timuthy) [inline pkg/operation/botanist/component/vpnseedserver/vpn_seed_server.go]: Nit: Create a helper function for `matchLabels` in `pkg/utils/gardener/networkpolicy.go`.

**Comment by rfranzke** [comment]
Interestingly, the `zero-down-time-validator-update` pod is failing even before the upgrade is started:

image

image

image

I can't reproduce this locally and have no idea why this happens or how it should be related to this PR (especially, since the PR is not even yet used) 👀

File: pkg/operation/botanist/component/kubeapiserver/deployment.go [hunk 4]
```diff
@@ -206,9 +207,10 @@ func (k *kubeAPIServer) reconcileDeployment(
 			Template: corev1.PodTemplateSpec{
 				ObjectMeta: metav1.ObjectMeta{
 					Labels: utils.MergeStringMaps(GetLabels(), map[string]string{
-						v1beta1constants.LabelNetworkPolicyToDNS:             v1beta1constants.LabelNetworkPolicyAllowed,
-						v1beta1constants.LabelNetworkPolicyToPublicNetworks:  v1beta1constants.LabelNetworkPolicyAllowed,
-						v1beta1constants.LabelNetworkPolicyToPrivateNetworks: v1beta1constants.LabelNetworkPolicyAllowed,
+						v1beta1constants.LabelNetworkPolicyToDNS:                                                                    v1beta1constants.LabelNetworkPolicyAllowed,
+						v1beta1constants.LabelNetworkPolicyToPublicNetworks:                                                         v1beta1constants.LabelNetworkPolicyAllowed,
+						v1beta1constants.LabelNetworkPolicyToPrivateNetworks:                                                        v1beta1constants.LabelNetworkPolicyAllowed,
+						gardenerutils.NetworkPolicyLabel(resourcemanagerconstants.ServiceName, resourcemanagerconstants.ServerPort): v1beta1constants.LabelNetworkPolicyAllowed,
```
Comment (timuthy) [inline pkg/operation/botanist/component/kubeapiserver/deployment.go]: As discussed: We should try to find a way to an additional `all-webhook-targets` policy/label that we can use here instead, similar to `all-scrape-targets`.

File: charts/seed-bootstrap/charts/loki/templates/allow-loki-network-policy.yaml [hunk 5]
```diff
@@ -41,13 +41,6 @@ spec:
         matchLabels:
           app: nginx-ingress
           component: controller
-    - namespaceSelector:
-        matchLabels:
-          role: kube-system
-      podSelector:
-        matchLabels:
-          app: nginx-ingress
-          component: controller
```
Comment (timuthy) [inline charts/seed-bootstrap/charts/loki/templates/allow-loki-network-policy.yaml:48]: Hm, I think this is due to previous setups where the Ingress deployment was managed via shoot addons (before https://github.com/gardener/gardener/pull/3131) and at least I couldn't spot that such a setup is not supported anymore when not using `ManagedSeed`s. We should at least add a breaking release note?!

**Review by timuthy — COMMENTED** [review]
Nice, I really like the new approach that brings in a lot more structure 🙂

File: pkg/operation/botanist/component/kubeapiserver/deployment.go (hunk 1 above)
Comment (rfranzke) [reply pkg/operation/botanist/component/kubeapiserver/deployment.go:751]: In this case, we already exit early and don't reach this code: https://github.com/gardener/gardener/blob/7265e2c1e9ca93e17ca2300d0ad73c6e0f62f9b2/pkg/operation/botanist/component/kubeapiserver/deployment.go#L722-L724

File: charts/seed-bootstrap/charts/loki/templates/allow-loki-network-policy.yaml (hunk 5 above)
Comment (rfranzke) [reply charts/seed-bootstrap/charts/loki/templates/allow-loki-network-policy.yaml:48]: LOL, that's a very good point. We have deprecated this scenario and the `.spec.dns.ingressDomain` field in the `Seed` with https://github.com/gardener/gardener/pull/3131, https://github.com/gardener/gardener/blob/7265e2c1e9ca93e17ca2300d0ad73c6e0f62f9b2/pkg/apis/core/v1beta1/types_seed.go#L144-L145

We should definitely drop this technical debt. I opened https://github.com/gardener/gardener/issues/7519 about it and added a release note to remind the people about it.

File: pkg/operation/botanist/component/kubeapiserver/deployment.go (hunk 4 above)
Comment (rfranzke) [reply pkg/operation/botanist/component/kubeapiserver/deployment.go]: Yes, let's pursue this separately. I think this PR already has a certain complexity, and this particular request is rather a nice addition/feature on to simplify things even further. OK?

Comment (timuthy) [reply pkg/operation/botanist/component/kubeapiserver/deployment.go]: Okay, fine for me.

**Review by timuthy — COMMENTED** [review]
Nice 🙂 
/lgtm

**Pull request merged**